content/lectures/lec00.md

```diff
@@ -8,3 +8,73 @@ draft: false
+
+We'll look at a few different package managers over the course of the semester. Node has one called NPM (Node Package Manager), Java has a package manager called Maven, and Python has a few offerings. Note that these package managers are all a little different, because they work with different languages that all have different nuances, this is why we can't reuse package managers across languages.
+
+The Python package manager we'll be using is called Poetry. Essentially, Poetry allows you to download certain Python libraries, then it creates a virtual python environment on your machine to run your code with the given libraries. So, why the virtual machine? The answer is that Python varies a lot from version to version (especially Python 2 compared to Python 3). The virtual environment ensures that you are running the same version of Python as your team of developers. It also allows us to ensure that the version of Python you run in development is the same as the version that runs in production.
+
+Now, let's get into how to actually use Poetry. First, make sure that you have Poetry installed on your machine, instructions for installation can be found [here](https://python-poetry.org/docs/).
+
+Once you have Poetry installed, let's create a new project:
+
+```
+# Create a new folder called poetry_demo
+$ mkdir poetry_demo
+# Enter the new folder
+$ cd poetry_demo
+$ poetry init
+```
+
+Now Poetry will give you lots of options for how to initialize your project, just hit enter for all of them (Poetry will go to the default setup which is fine for our purposes). Once you've finished, you'll see that there is a new file `pyproject.toml` in the directory, this is the file that stores the information we just initialized.
+
+Next, let's add a dependency:
+
+```
+$ poetry add redis
+Creating virtualenv poetry-demo-KkU142w6-py3.9 in /Users/airbenderang/Library/Caches/pypoetry/virtualenvs
+Using version ^1.19.5 for numpy
+
+Updating dependencies
+Resolving dependencies... (39.8s)
+
+Writing lock file
+
+Package operations: 1 install, 0 updates, 0 removals
+
+  • Installing numpy (1.19.5)
+```
+
+Now, Poetry actually does two things here. It downloads Numpy, but before that it actually creates a virtual environment which we are going to use to run our Python code. Now, for our future Poetry projects, they will use the same virtual environment unless we want to configure Poetry to create a virtual environment with a deprecated version of Python. Again, you will see a new file in your directory, this is the `poetry.lock` file that doesn't make much sense to humans but tracks which packages your program needs, and the version number of those packages.
```

Comment:
this is not true. each venv the poetry creates is unique with respect to the project

Reply:
TIL. Thank you!